internal/controller/controllers/agent_controller.go

```diff
@@ -204,6 +209,95 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, &h.Host, h.ClusterID, nil, false)
 }
 
+func (r *AgentReconciler) shouldApproveMoreCSRs(node *corev1.Node) bool {
+	return !isNodeReady(node)
+}
+
+// Validate that the CSR can be approved
+func (r *AgentReconciler) shouldApproveCSR(csr *certificatesv1.CertificateSigningRequest, agent *aiv1beta1.Agent, validateNodeCsr nodeCsrValidator) (bool, error) {
+	x509CSR, err := getX509ParsedRequest(csr)
+	if err != nil {
+		return false, err
+	}
+	if !isCsrAssociatedWithAgent(x509CSR, agent) {
+		return false, nil
+	}
+
+	if !isCsrInTimeSpan(agent, csr) {
+		return false, errors.Errorf("CSR %s of agent %s/%s: creation timestamp %v is not in required time span",
+			csr.Name, agent.Namespace, agent.Name, csr.CreationTimestamp.Time)
+	}
+	return validateNodeCsr(agent, csr, x509CSR)
+}
+
+func (r *AgentReconciler) approveAIHostsCSRs(clients SpokeK8sClient, agent *aiv1beta1.Agent, validateNodeCsr nodeCsrValidator) {
+	csrList, err := clients.ListCsrs()
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to get CSRs for agent %s/%s", agent.Namespace, agent.Name)
+		return
+	}
+	for i := range csrList.Items {
+		csr := &csrList.Items[i]
+		if !isCsrApproved(csr) {
+			shouldApprove, err := r.shouldApproveCSR(csr, agent, validateNodeCsr)
+			if err != nil || !shouldApprove {
+				if err != nil {
+					r.Log.WithError(err).Errorf("Failed checking if CSR %s should be approved for agent %s/%s",
+						csr.Name, agent.Namespace, agent.Name)
+				}
+				continue
+			}
+			if err = clients.ApproveCsr(csr); err != nil {
+				r.Log.WithError(err).Errorf("Failed to approve CSR %s for agent %s/%s", csr.Name, agent.Namespace, agent.Name)
+				continue
+			}
+		}
+	}
+}
+
+// Attempt to approve CSRs for agent. If already approved then the node will be marked as done
+// requeue means that approval will be attempted again
+func (r *AgentReconciler) tryApproveDay2CSRs(ctx context.Context, agent *aiv1beta1.Agent) bool {
+	r.Log.Infof("Approving CSRs for agent %s/%s", agent.Namespace, agent.Name)
+	namespacedName := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      fmt.Sprintf(adminKubeConfigStringTemplate, agent.Spec.ClusterDeploymentName.Name),
+	}
+
+	secret, err := getSecret(ctx, r.Client, r.APIReader, namespacedName)
+	if err != nil {
+		r.Log.WithError(err).Errorf("Agent %s/%s: Failed to get secret", agent.Namespace, agent.Name)
+		return false
+	}
+	clients, err := r.SpokeK8sClientFactory.Create(secret)
+	if err != nil {
+		r.Log.WithError(err).Errorf("Agent %s/%s: Failed to create spoke client", agent.Namespace, agent.Name)
+		return false
+	}
+	hostname := getAgentHostname(agent)
+	var (
+		validateNodeCsr       nodeCsrValidator
+		shouldApproveMoreCSRs bool
+	)
+	node, err := clients.GetNode(hostname)
```

Comment:
Let's add a `TODO` so that we can check the proper hostname for the IPv6 case. TL;DR: In IPv6, the node will have a full FQDN as name rather than the hostname